### PR TITLE
nillable deploy resource + nillable LWRP args

### PR DESCRIPTION
**Pull request by lamont-granquist**
introduced nillable_set_or_return so that we can actually set a nil value
on a resource from the DSL.

this fixes the case where you want to do (and we document) setting
symlink_before_migrate to nil on the deploy resource.

to implement this better also moved up the attribute DSL method to the
base Chef::Resource class.

all LWRP resources are now nillable as well.

### CODE REVIEW

**Comment by lamont-granquist** [comment]
@chef/client-core @chef/client-engineers this one should be pretty cool

File: lib/chef/mixin/params_validate.rb [hunk 1]
```diff
@@ -81,34 +81,58 @@ def lazy(&block)
...
-      def set_or_return(symbol, arg, validation)
+      NULL_ARG = Object.new
+
+      def nillable_set_or_return(symbol, arg, validation)
         iv_symbol = "@#{symbol.to_s}".to_sym
-        if arg == nil && self.instance_variable_defined?(iv_symbol) == true
-          ivar = self.instance_variable_get(iv_symbol)
-          if(ivar.is_a?(DelayedEvaluator))
-            validate({ symbol => ivar.call }, { symbol => validation })[symbol]
+        if arg == NULL_ARG
```
Comment (danielsdeleo) [inline lib/chef/mixin/params_validate.rb]: for maximum ruby pedantry you might want to do `NULL_ARG.equal?(arg)` instead. `arg == other` behavior is defined by how `arg` implements `==` which could possibly (not super likely) be insane. Also this is a rare case where using object equality is exactly what you want (even though for Object, `==` and `eql?` are just aliases for `equal?`).

Comment (lamont-granquist) [inline lib/chef/mixin/params_validate.rb]: Yep, both good points, fixing...

**Comment by danielsdeleo** [comment]
Pedantry aside, :+1:

**Comment by tyler-ball** [comment]
:+1:

**Pull request merged**